### Brief Overview

* In this assignment, you will practice some data wrangling functionality commonly used in real-world projects.

* This assignment is designed around two dataset

* The Medicaid Data per State  

* IRS Statistics of Income (SOI) dataset


* The final product of this assignment will be a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  * Which medications account for the bulk of a state's spending   
  * Which medications are prescribed more frequently in one state than another.
  * etc.

* The following URLs provide the datasets required for this assignment:
  * [medicaid_data.csv](https://www.dropbox.com/s/u6ctjqxnk0wxpbk/medicaid_data.csv?dl=1)
  * [medicaid_enrollment.tsv](https://www.dropbox.com/s/969mohzhpu78r10/medicaid_enrollment.tsv?dl=1)
  * [tax_data.csv](https://www.dropbox.com/s/zm37nu7vnirha4m/tax_data.csv?dl=1)


 ### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame` or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data ` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

  - `info()` is a method `tax_data`
  
```python
   
    tax_data.info()
```

  or

  - `shape` is a property of `tax_data`


```python
tax_data.shape
```

- Similarly, if you are asked to find to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using set() or len() are not acceptable for the following reasons:

- The solution below uses `set()` and `len()`, which are not `tax_data` methods

```python
len(set(tax_data["STATEFIPS"]))
```

- The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of uniques entries using `len()` which is not `tax_data` method

```python
len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (`tax_data['STATEFIPS'])

```python
tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity 

- For instance, to identify whether a value is part of the index, write the following:

```python
tax_data.index.contains(99999)
```

- Rather than:

```python
9999 in tax_data.index
```

- You can only import `pandas` and `numpy`

- If you are not explicitly asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external modules will be accepted.

- __Important__: Provide the exact statement(s) used to answer each question

- Unless otherwise specified, each cell can contain multiple lines of code

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, assuming you are using Jupyter Notebooks, I suggest you use `.SHIFT+TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` twice to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora of functionality available through the `pandas` package. 


### 1. Loading and exploring the tax data

Import Pandas and numpy using their appropriate alias here


In [1]:
# WRITE YOUR CODE HERE

import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (`tax_data.csv`) into a `DataFrame` called `tax_data`. This file can be downloaded from the `URL` provided above.

* This file `tax_data.csv` was preprocessed from the original file I obtained at the following URL:
  * [15zpallagi.csv](https://www.irs.gov/pub/irs-soi/15zpallagi.csv}

In [2]:
## WRITE YOUR CODE HERE 
tax_data = pd.read_csv("dataset/tax_data.csv")

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [3]:
## WRITE YOUR CODE HERE 
tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 131 columns]

*  Modify `tax_data` to uppercase all the header names. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can use `tax_data.columns`, which returns a `Series` of the column names.

* The resulting column name should look as follows:

```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

* Standardizing column names is a good way to avoid having to guess whether the column header is in lowercase or uppercase.

In [4]:
## WRITE YOUR CODE HERE 
tax_data.columns = tax_data.columns.str.upper()
tax_data.columns

Index(['STATEFIPS', 'STATE', 'ZIPCODE', 'AGI_STUB', 'N1', 'MARS1', 'MARS2',
       'MARS4', 'PREP', 'N2',
       ...
       'N10300', 'A10300', 'N85530', 'A85530', 'N85300', 'A85300', 'N11901',
       'A11901', 'N11902', 'A11902'],
      dtype='object', length=131)

* What is the total number of entries (also called observations or instances) in `tax_data` dataset?
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [5]:
## WRITE YOUR CODE HERE
print("Number of rows:", tax_data.shape[0])
print("Number of columns:", tax_data.shape[1])

Number of rows: 166698
Number of columns: 131


* If `STATEFIPS` is the header label of the first column in `tax_data`, write code to print the title of the 32nd column.  
  * You should use a single python expression
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [6]:
## WRITE YOUR CODE HERE 
tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the first column label (index 0), what is the index of the column labeled `N10300`?
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [7]:
## WRITE YOUR CODE HERE
tax_data.columns.get_loc("N10300")

121

- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties

- The result should look like the following ( '...' represents remaining data that is not shown here)
```python

        ZIPCODE
STATE	
TX	     9714
NY	     9238
CA	     8908
PA	     8208
IL	     7386
...
```

* The above indicates that there are 9,714 unique zip codes in TX, 9,238 in NY, etc.


In [10]:
## WRITE YOUR CODE HERE 
unique_zip = tax_data.groupby('STATE')['ZIPCODE'].nunique()
unique_zip = unique_zip.reset_index()
sorted_unique_zip = unique_zip.sort_values(by='ZIPCODE', ascending=False)
sorted_unique_zip = sorted_unique_zip.reset_index(drop=True)
sorted_unique_zip

STATE  ZIPCODE
0     TX     1619
1     NY     1540
2     CA     1485
3     PA     1368
4     IL     1231
5     OH      998
6     FL      920
7     MI      892
8     MO      890
9     IA      827
10    VA      793
11    MN      792
12    NC      725
13    WI      714
14    IN      674
15    GA      667
16    KY      656
17    KS      598
18    TN      590
19    AL      577
20    NJ      547
21    OK      547
22    WV      512
23    WA      497
24    AR      490
25    NE      486
26    MA      482
27    LA      453
28    MD      405
29    CO      395
30    SC      376
31    MS      370
32    ME      370
33    OR      356
34    AZ      295
35    ND      287
36    SD      287
37    CT      264
38    VT      238
39    NH      232
40    MT      225
41    ID      213
42    NM      209
43    UT      186
44    NV      128
45    WY      109
46    RI       71
47    HI       60
48    DE       57
49    AK       56
50    DC       24

* Identify the position of `HI` in the list of zip code counts per state generated in the previous question
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [11]:
## WRITE YOUR CODE HERE 
sorted_unique_zip.loc[sorted_unique_zip['STATE'] == 'HI']

STATE  ZIPCODE
47    HI       60

### 2. Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [12]:
## WRITE YOUR CODE HERE 
invalid_zip = tax_data[tax_data['ZIPCODE'] == 0]
nb_invalid_zip = invalid_zip.shape[0]
nb_invalid_zip

306

* Run the line in the code cell below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
* If the code cell below returns an error, then change your answer above so that the value returned is a number.

In [13]:
assert(type(nb_invalid_zip) == int)

* Remove from `tax_data` all the instances where the zip code is `0` and save the resulting `DataFrame` to a variable named `tax_data_valid_zip`

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [27]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip = tax_data[tax_data['ZIPCODE'] != 0]

* Run the line below to confirm that the operation worked as expected

  * The assertion below verifies that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [28]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [19]:
## WRITE YOUR CODE HERE 
nb_missing_values = tax_data_valid_zip.shape[0]-tax_data_valid_zip.dropna().shape[0]
nb_missing_values

199

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values (`NaN`)

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [23]:
## WRITE YOUR CODE HERE
tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna()
tax_data_valid_zip_cleaned = tax_data_valid_zip_cleaned.reset_index()
tax_data_valid_zip_cleaned.shape[0]

166193

* Run the line below to confirm that the operation worked as expected. The assertion below tests that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [24]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code
* The function `compute_percentile_zipcode` below computes the percentile income per zip code
* By default `percentile=0.5`, i.e., the function computes the median
* Read the code and make sure you understand what it does before moving on to the next question

In [29]:
def compute_percentile_zip(df_zip, percentile=0.5):     
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zip code in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * You can use any of the submethods discussed in the Split-Apply-Combine paradigm.
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`

* The resulting `Series` should look like the following ( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [30]:
## WRITE YOUR CODE HERE 
computed_order = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip)
zip_rev_all = computed_order.sort_values(ascending=False)

* What are the three zip codes with the most significant 65th percentile value for income?
 * I.e., the top three zip codes in your sorted listed

In [31]:
## WRITE YOUR CODE HERE 
zip_rev_all.head(3)

ZIPCODE
33109.0    3954.114286
77010.0    1033.991667
31561.0     883.925000
dtype: float64

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file can be downloaded from the URL provided above. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (e.g., less than 6-8 GB)


In [33]:
## WRITE YOUR CODE HERE
medicaid_data = pd.read_csv("dataset/medicaid_data.csv")
medicaid_data.head()

Utilization Type State          NDC Product Name  Units Reimbursed  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   Number of Prescriptions  Total Amount Reimbursed  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   Medicaid Amount Reimbursed  Non Medicaid Amount Reimbursed  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               Location  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)

* Modify `medicaid_data` to capitalize all column names 

  * If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [35]:
## WRITE YOUR CODE HERE 
medicaid_data.columns = medicaid_data.columns.str.upper()
medicaid_data.columns

Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')

* Familiarize yourself with the data
  * the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  * The remaining column names are self-explanatory
* Explore the number of lines and columns in the data
* Check that your column headers are capitalized

In [36]:
## WRITE YOUR CODE HERE
print("Number of lines:", medicaid_data.shape[0])
print("Number of columns:", medicaid_data.shape[1])
print("Columns:", medicaid_data.columns)

Number of lines: 1695546
Number of columns: 10
Columns: Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')


* If you explore the `LOCATION` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical.
* Are there any states that have more than one value for `LOCATION`. 
  * This question can be answered by using aggregation and the Split-Apply-Combine paradigm
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [38]:
## WRITE YOUR CODE HERE 
hawaii = medicaid_data.groupby('STATE').get_group('HI')
hawaii.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
154             MCOU    HI  16714034804   DASETTA 1-             448.0   
294             MCOU    HI  49348002972    SM TRIPLE             821.6   
318             MCOU    HI    378043301    BUPROPION            5461.0   
397             MCOU    HI  49348004537   SM ALLERGY            5016.0   
557             MCOU    HI     54007928   BALSALAZID            3630.0   

     NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
154                     16.0                   267.07   
294                     27.0                   143.34   
318                     97.0                  2206.56   
397                     35.0                   105.57   
557                     14.0                  3204.75   

     MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
154                      267.07                             0.0   
294                      143.34                             0.0   
318                     2206.56                             0.0   
397                      105.57                             0.0   
557                     3204.75                             0.0   

                 LOCATION  
154  (21.1098, -157.5311)  
294  (21.1098, -157.5311)  
318  (21.1098, -157.5311)  
397  (21.1098, -157.5311)  
557  (21.1098, -157.5311)

In [39]:
group_by_state = medicaid_data.groupby('STATE')
result = group_by_state.nunique()
result
## There is no state that has more than one value for LOCATION

UTILIZATION TYPE    NDC  PRODUCT NAME  UNITS REIMBURSED  \
STATE                                                            
AK                    1   5146           880              6113   
AL                    1   9071          1983             12059   
AR                    1   7236           950              9743   
AZ                    2   9929          1408             14097   
CA                    2  17411          3012             38151   
CO                    2   9401          1433             13350   
CT                    1   9596          2652             13044   
DC                    2   4646           932              6473   
DE                    2   5156          1397              6450   
FL                    2  13153          1787             22976   
GA                    2  11252          2359             18974   
HI                    2   5539          1172              6732   
IA                    2   7759          1811             10446   
ID                    1   5827           901              7579   
IL                    2  12636          2624             23152   
IN                    2  11344          2561             18598   
KS                    2   9026          2122              9820   
KY                    2  13555          1705             19830   
LA                    2  11442          1524             16847   
MA                    2  10189          1984             17664   
MD                    2  11797          1847             16777   
ME                    1   5654           989              7609   
MI                    2  13008          1606             22218   
MN                    2  11030          1536             18504   
MO                    1  11951          1591             16349   
MS                    2   7818          1222             10900   
MT                    1   6386          1187              7793   
NC                    1  11887          1577             17506   
ND                    2   3309           743              4574   
NE                    2   6834          1761              8515   
NH                    2   3996           756              5241   
NJ                    2  12974          2773             16442   
NM                    2   7669          1271             10273   
NV                    2   8055          1947             12631   
NY                    2  17858          3137             35213   
OH                    2  16010          3426             29174   
OK                    1   7802          1790             10748   
OR                    2   9678          2238             14674   
PA                    2  15341          3251             24560   
RI                    2   5331          1464              7678   
SC                    2   8075          1202             11679   
SD                    1   3924          1088              4388   
TN                    2   9693          1261             14389   
TX                    2  12906          2387             21944   
UT                    2   6217          1571              8360   
VA                    2  10817          2946             16696   
VT                    1   4402          1241              6127   
WA                    2  10904          2363             17417   
WI                    2  10227          2427             15768   
WV                    2   9398          1456             14342   
WY                    1   2751           794              3401   
XX                    2  30472          6367             89897   

       NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE                                                     
AK                         625                    12006   
AL                        1817                    23075   
AR                        1455                    17882   
AZ                        2707                    26651   
CA                        6132                    74516   
CO                        189

* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state.  That is, we need to compare per capita. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.

 
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescriptions in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).


https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained [here](https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html)

* A parsed/processed version (`medicaid_enrollment.tsv`) can be downloaded from the URL provided above. Use `pandas` to load the `medicaid_enrollment.tsv` file into DataFrame named `medicaid_enrollment`

In [40]:
## WRITE YOUR CODE HERE 
medicaid_enrollment = pd.read_csv('dataset/medicaid_enrollment.tsv', sep='\t')
medicaid_enrollment

STATE Total Medicaid Enrollees
0                    Alabama                1,050,989
1                     Alaska                  164,783
2             American Samoa                      NaN
3                    Arizona                1,740,520
4                   Arkansas                  762,166
5                 California               13,096,861
6                   Colorado                1,264,600
7                Connecticut                  746,119
8                   Delaware                  227,909
9       District of Columbia                  271,428
10                   Florida                3,808,334
11                   Georgia                1,990,810
12                      Guam                      n/a
13                    Hawaii                  340,513
14                     Idaho                  283,355
15                  Illinois                3,269,999
16                   Indiana                1,295,358
17                      Iowa                  618,505
18                    Kansas                  403,844
19                  Kentucky                1,284,193
20                 Louisiana                1,402,212
21                     Maine                  288,324
22                  Maryland                1,271,445
23             Massachusetts                1,829,618
24                  Michigan                3,947,031
25                 Minnesota                1,052,521
26               Mississippi                  740,937
27                  Missouri                  944,257
28                   Montana                  139,950
29                  Nebraska                  239,463
30                    Nevada                  588,304
31             New Hampshire                  186,399
32                New Jersey                1,705,594
33                New Mexico                  826,155
34                  New York                6,281,038
35            North Carolina                1,965,805
36              North Dakota                   86,250
37  Northern Mariana Islands                      NaN
38                      Ohio                3,060,446
39                  Oklahoma                  829,561
40                    Oregon                1,123,913
41              Pennsylvania                2,569,232
42               Puerto Rico                1,458,819
43              Rhode Island                  308,521
44            South Carolina                1,233,430
45              South Dakota                  124,497
46                 Tennessee                1,562,745
47                     Texas                4,273,982
48                      Utah                  293,867
49                   Vermont                  206,469
50            Virgin Islands                      NaN
51                  Virginia                1,092,225
52                Washington                1,771,679
53             West Virginia                  545,748
54                 Wisconsin                1,209,714
55                   Wyoming                   66,532

* Modify `medicaid_enrollment` `DataFrame` to capitalize all the column names 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by manually uppercasing the columns

In [41]:
## WRITE YOUR CODE HERE 
medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()
medicaid_enrollment.columns

Index(['STATE', 'TOTAL MEDICAID ENROLLEES'], dtype='object')

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`
* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing value means in this context

In [42]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
medicaid_enrollment_cleaned = medicaid_enrollment_cleaned[medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].str.contains('n/a') == False]
medicaid_enrollment_cleaned

STATE TOTAL MEDICAID ENROLLEES
0                Alabama                1,050,989
1                 Alaska                  164,783
3                Arizona                1,740,520
4               Arkansas                  762,166
5             California               13,096,861
6               Colorado                1,264,600
7            Connecticut                  746,119
8               Delaware                  227,909
9   District of Columbia                  271,428
10               Florida                3,808,334
11               Georgia                1,990,810
13                Hawaii                  340,513
14                 Idaho                  283,355
15              Illinois                3,269,999
16               Indiana                1,295,358
17                  Iowa                  618,505
18                Kansas                  403,844
19              Kentucky                1,284,193
20             Louisiana                1,402,212
21                 Maine                  288,324
22              Maryland                1,271,445
23         Massachusetts                1,829,618
24              Michigan                3,947,031
25             Minnesota                1,052,521
26           Mississippi                  740,937
27              Missouri                  944,257
28               Montana                  139,950
29              Nebraska                  239,463
30                Nevada                  588,304
31         New Hampshire                  186,399
32            New Jersey                1,705,594
33            New Mexico                  826,155
34              New York                6,281,038
35        North Carolina                1,965,805
36          North Dakota                   86,250
38                  Ohio                3,060,446
39              Oklahoma                  829,561
40                Oregon                1,123,913
41          Pennsylvania                2,569,232
42           Puerto Rico                1,458,819
43          Rhode Island                  308,521
44        South Carolina                1,233,430
45          South Dakota                  124,497
46             Tennessee                1,562,745
47                 Texas                4,273,982
48                  Utah                  293,867
49               Vermont                  206,469
51              Virginia                1,092,225
52            Washington                1,771,679
53         West Virginia                  545,748
54             Wisconsin                1,209,714
55               Wyoming                   66,532

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization



In [43]:
## WRITE YOUR CODE HERE
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'] = medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].str.replace(',','').astype(int)

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and make sure that the data type is `int`

In [44]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].dtype

dtype('int64')

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

* We can use the shared State information across both tables to associate both tables (SQL JOIN).

* However, `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name

  * We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

* Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.

  * A brief description of what the code does is included in the comments
  
* Note that the code below may throw an error if `lxml` is not installed on your machine. You should install any Python packages that Python complains about before proceeding.

In [45]:

import requests

url = 'https://www.50states.com/abbreviations.htm'
header = {
  "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
  "X-Requested-With": "XMLHttpRequest"
}

r = requests.get(url, headers=header)

tables = pd.read_html(r.text)


# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

US STATE POSTAL ABBREVIATION STANDARD ABBREVIATION
0     Alabama                  AL                  Ala.
1      Alaska                  AK                Alaska
2     Arizona                  AZ                 Ariz.
3    Arkansas                  AR                  Ark.
4  California                  CA                Calif.

* Change the the `DataFrame`'s headers from ['US STATE',	'POSTAL ABBREVIATION', 	'STANDARD ABBREVIATION'] to ['US STATE', 'ABBREVIATION', `STD ABBREVIATION`]
  * You can hard code this operation

In [48]:
## WRITE YOUR CODE HERE 
Codes_abbreviations = Codes_abbreviations.rename(columns={'POSTAL ABBREVIATION':'ABBREVIAION','STANDARD ABBREVIATION':'STD ABBREVIATION'})
Codes_abbreviations.head()

US STATE ABBREVIAION STD ABBREVIATION
0     Alabama          AL             Ala.
1      Alaska          AK           Alaska
2     Arizona          AZ            Ariz.
3    Arkansas          AR             Ark.
4  California          CA           Calif.

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 

* Save the results to a variable named `medicaid_enrollment_cleaned_with_zip`

* `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online (e.g., `https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`) and in a notebook under `Data Wrangling - Advanced` module. See for instance:


* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [49]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.merge(Codes_abbreviations, how='left', left_on='STATE', right_on='US STATE')
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop(['US STATE','STD ABBREVIATION'], axis=1)
medicaid_enrollment_cleaned_with_zip.head()

STATE  TOTAL MEDICAID ENROLLEES ABBREVIAION
0     Alabama                   1050989          AL
1      Alaska                    164783          AK
2     Arizona                   1740520          AZ
3    Arkansas                    762166          AR
4  California                  13096861          CA

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`

  * Remove the column and make sure the data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):



```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....

```

In [50]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop('STATE', axis=1)
medicaid_enrollment_cleaned_with_zip.head()

TOTAL MEDICAID ENROLLEES ABBREVIAION
0                   1050989          AL
1                    164783          AK
2                   1740520          AZ
3                    762166          AR
4                  13096861          CA

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not relevant. This answer uses the same approach as the one used to `merge` the tables above. 

In [51]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.rename(columns={'ABBREVIAION':'STATE'})
medicaid_data_w_enrollments = pd.merge(medicaid_data, medicaid_enrollment_cleaned_with_zip)
medicaid_data_w_enrollments.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    PA    456060010   TEFLARO 60             224.0   
2             MCOU    PA    115692201   DIVALPROEX            1626.0   
3             FFSU    PA     54455025    METHOTREX             450.0   
4             MCOU    PA  50458014090     INVOKANA            3850.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     36.0                 35752.31   
2                     20.0                  2318.55   
3                     22.0                   565.29   
4                    111.0                 52655.13   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    35752.31                            0.00   
2                     2287.80                           30.75   
3                      516.23                           49.06   
4                    50641.22                         2013.91   

             LOCATION  TOTAL MEDICAID ENROLLEES  
0  (40.5773, -77.264)                   2569232  
1  (40.5773, -77.264)                   2569232  
2  (40.5773, -77.264)                   2569232  
3  (40.5773, -77.264)                   2569232  
4  (40.5773, -77.264)                   2569232

- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [52]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments = medicaid_data_w_enrollments.dropna(subset=['STATE','PRODUCT NAME'])
medicaid_data_w_enrollments.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    PA    456060010   TEFLARO 60             224.0   
2             MCOU    PA    115692201   DIVALPROEX            1626.0   
3             FFSU    PA     54455025    METHOTREX             450.0   
4             MCOU    PA  50458014090     INVOKANA            3850.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     36.0                 35752.31   
2                     20.0                  2318.55   
3                     22.0                   565.29   
4                    111.0                 52655.13   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    35752.31                            0.00   
2                     2287.80                           30.75   
3                      516.23                           49.06   
4                    50641.22                         2013.91   

             LOCATION  TOTAL MEDICAID ENROLLEES  
0  (40.5773, -77.264)                   2569232  
1  (40.5773, -77.264)                   2569232  
2  (40.5773, -77.264)                   2569232  
3  (40.5773, -77.264)                   2569232  
4  (40.5773, -77.264)                   2569232

- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [59]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch = medicaid_data_w_enrollments.set_index(['STATE','NDC'])
medicaid_data_w_enrollments_hierarch = medicaid_data_w_enrollments_hierarch.sort_index()
medicaid_data_w_enrollments_hierarch

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
AK    2143380                 FFSU    TRULICITY              57.0   
      2143380                 FFSU    TRULICITY              58.0   
      2143380                 FFSU    TRULICITY              96.0   
      2143480                 FFSU    TRULICITY              46.0   
      2143480                 FFSU    TRULICITY              56.0   
...                            ...          ...               ...   
WY    76385010401             FFSU   Benztropin            1020.0   
      76385010401             FFSU   Benztropin             780.0   
      76385010501             FFSU   Benztropin            1065.0   
      76439030910             FFSU   HYOSCYAMIN            1214.0   
      76439030910             FFSU   HYOSCYAMIN            1184.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
AK    2143380                         28.0                 17493.58   
      2143380                         28.0                 18439.17   
      2143380                         48.0                 31466.12   
      2143480                         23.0                 14096.17   
      2143480                         28.0                 17996.75   
...                                    ...                      ...   
WY    76385010401                     21.0                   299.72   
      76385010401                     12.0                   148.80   
      76385010501                     22.0                   334.64   
      76439030910                     20.0                   467.07   
      76439030910                     22.0                   485.54   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
AK    2143380                        17493.58                             0.0   
      2143380                        18439.17                             0.0   
      2143380                        31466.12                             0.0   
      2143480                        14096.17                             0.0   
      2143480                        17996.75                             0.0   
...                                       ...                             ...   
WY    76385010401                      299.72                             0.0   
      76385010401                      148.80                             0.0   
      76385010501                      334.64                             0.0   
      76439030910                      467.07                             0.0   
      76439030910                      485.54                             0.0   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
AK    2143380       (61.385, -152.2683)                    164783  
      2143380       (61.385, -152.2683)                    164783  
      2143380       (61.385, -152.2683)                    164783  
      2143480       (61.385, -152.2683)                    164783  
      2143480       (61.385, -152.2683)                    164783  
...                                 ...                       ...  
WY    76385010401  (42.7475, -107.2085)                     66532  
      76385010401  (42.7475, -107.2085)                     66532  
      76385010501  (42.7475, -107.2085)                     66532  
      76439030910  (42.7475, -107.2085)                     66532  
      76439030910  (42.7475, -107.2085)                     66532  

[1526193 rows x 9 columns]


* Write a single Pandas expression to print all the lines containing `NDC` `61958180101` in the Pennsylvania ("PA")

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [60]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch.loc[('PA', 61958180101)]

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            2604.0   
      61958180101             FFSU   Harvoni  (            1008.0   
      61958180101             MCOU   Harvoni  (            1932.0   
      61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            3220.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    61958180101                     33.0               1017644.54   
      61958180101                     93.0               2952891.16   
      61958180101                     36.0               1107725.22   
      61958180101                     69.0               2178602.30   
      61958180101                     33.0               1015383.60   
      61958180101                    115.0               3599192.35   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    61958180101                   985269.84                        32374.70   
      61958180101                  2860100.54                        92790.62   
      61958180101                  1107725.22                            0.00   
      61958180101                  2118612.30                        59990.00   
      61958180101                  1015383.60                            0.00   
      61958180101                  3482188.68                       117003.67   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232

#### Computing the normalized `UNITS REIMBURSED` per drug 



* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.



- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612



```python

total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 

```



- And the numebr of Medicaid Enrollees in "PA" is 2,569,232





```python

total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]

```

- Therefore, the UNITS REIMBURSED per enrollee for "HARVONI" is  0.00413041718303





```python

print(total_amount_reimbursed/total_enrollees_PA)

```



- Rather than work directly with the ratios, we are going to compute and work with their logarithm (np.log2) instead.



  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures



- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`





- Your final result should be a `Series` that looks like the following ( `...` represents data that is not shown here):
  * Note that negative values are the result of the `log` transformation 

```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...
AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...
MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626
...
```



In [64]:
## WRITE YOUR CODE HERE
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.groupby(['STATE', 'NDC'])['UNITS REIMBURSED'].sum()
total_enrollees_PA = medicaid_data_w_enrollments_hierarch.groupby(['STATE', 'NDC'])['TOTAL MEDICAID ENROLLEES'].unique()
units_reimbursed =  total_amount_reimbursed / total_enrollees_PA
units_reimbursed = units_reimbursed.astype(float)
medicaid_reimbursement_per_enrollee = np.log2(units_reimbursed)
medicaid_reimbursement_per_enrollee

STATE  NDC        
AK     2143380        -9.609109
       2143480       -10.008280
       2322730        -6.109830
       2322830        -4.444321
       2322930        -3.855995
                        ...    
WY     76329336901    -9.773833
       76385010301    -6.146779
       76385010401    -4.707744
       76385010501    -5.965123
       76439030910    -4.794145
Length: 468370, dtype: float64

- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`
- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - `index` is the two-letter state symbol 
  - column names are the `NDC` codes 
- The `DataFrame` should be formatted as in the image below
  - Hint: simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [65]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc = medicaid_reimbursement_per_enrollee.unstack(level=-1)
medicaid_norm_ndc.head()

NDC    2143380      2143480      2144511      2144527      2197590      \
STATE                                                                    
AK       -9.609109   -10.008280          NaN          NaN          NaN   
AL       -9.940595    -9.805485          NaN          NaN          NaN   
AR      -12.985157   -12.657103          NaN          NaN          NaN   
AZ      -11.533870   -10.821194          NaN          NaN          NaN   
CA      -10.926863   -10.698372   -16.127018          NaN    -9.464364   

NDC    2322730      2322830      2322930      2323030      2323130      ...  \
STATE                                                                   ...   
AK       -6.109830    -4.444321    -3.855995          NaN          NaN  ...   
AL       -5.336260    -4.243688    -3.645575          NaN          NaN  ...   
AR       -4.720964    -2.886635    -2.617579          NaN          NaN  ...   
AZ       -5.312806    -3.998085    -3.401510          NaN          NaN  ...   
CA       -7.429936    -6.286911    -5.864352          NaN          NaN  ...   

NDC    76439035930  76439035990  76439036290  76439036390  76439036490  \
STATE                                                                    
AK             NaN          NaN          NaN          NaN          NaN   
AL             NaN          NaN          NaN          NaN    -8.518493   
AR             NaN          NaN          NaN          NaN          NaN   
AZ      -11.824196          NaN          NaN          NaN          NaN   
CA             NaN          NaN          NaN          NaN          NaN   

NDC    76439036590  99207013070  99207024005  99207026012  99207046330  
STATE                                                                   
AK             NaN          NaN          NaN          NaN          NaN  
AL             NaN          NaN          NaN          NaN          NaN  
AR             NaN          NaN          NaN          NaN          NaN  
AZ             NaN          NaN          NaN          NaN          NaN  
CA             NaN   -11.952415          NaN          NaN   -15.150865  

[5 rows x 23883 columns]

#### Exploring the Data 

* What is the drug with the highest log-normalized ratio in Hawaii?


In [66]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc.loc['HI'].sort_values(ascending=False)

NDC
43386009019    10.511181
43386006019     3.448037
10572010001     2.599661
116200116       2.284969
62175044601     2.195764
                 ...    
76439036590          NaN
99207013070          NaN
99207024005          NaN
99207026012          NaN
99207046330          NaN
Name: HI, Length: 23883, dtype: float64

* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)

* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii? (Not graded)

In [70]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch.loc[('HI', 43386009019)]

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
HI    43386009019             MCOU   GAVILYTE-G          472000.0   
      43386009019             MCOU   GAVILYTE-G          376012.0   
      43386009019             MCOU   GAVILYTE-G       496104006.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
HI    43386009019                    118.0                  1754.68   
      43386009019                    105.0                  1462.60   
      43386009019                     63.0                   847.56   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
HI    43386009019                     1754.68                            0.00   
      43386009019                     1462.60                            0.00   
      43386009019                      839.61                            7.95   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
HI    43386009019  (21.1098, -157.5311)                    340513  
      43386009019  (21.1098, -157.5311)                    340513  
      43386009019  (21.1098, -157.5311)                    340513

In [71]:
medicaid_data_w_enrollments_hierarch.loc[('MA', 43386009019)]

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
MA    43386009019             FFSU   GAVILYTE-G             461.0   
      43386009019             MCOU   GAVILYTE-G             486.0   
      43386009019             FFSU   GAVILYTE-G            1086.0   
      43386009019             FFSU   GAVILYTE-G             676.0   
      43386009019             MCOU   GAVILYTE-G             753.0   
      43386009019             MCOU   GAVILYTE-G            1971.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
MA    43386009019                    294.0                  7135.59   
      43386009019                    471.0                  4587.04   
      43386009019                    479.0                  9185.81   
      43386009019                    499.0                 11456.60   
      43386009019                    727.0                  7264.42   
      43386009019                    925.0                 12118.86   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
MA    43386009019                     5720.40                         1415.19   
      43386009019                     4587.04                            0.00   
      43386009019                     9185.81                            0.00   
      43386009019                     9617.89                         1838.71   
      43386009019                     7264.42                            0.00   
      43386009019                     9820.33                         2298.53   

                              LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                         
MA    43386009019  (42.2373, -71.5314)                   1829618  
      43386009019  (42.2373, -71.5314)                   1829618  
      43386009019  (42.2373, -71.5314)                   1829618  
      43386009019  (42.2373, -71.5314)                   1829618  
      43386009019  (42.2373, -71.5314)                   1829618  
      43386009019  (42.2373, -71.5314)                   1829618

In [72]:
medicaid_data_w_enrollments_hierarch.loc[('FL', 43386009019)]

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
FL    43386009019             MCOU   GAVILYTE-G          596000.0   
      43386009019             MCOU   GAVILYTE-G          676000.0   
      43386009019             MCOU   GAVILYTE-G          576000.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
FL    43386009019                    149.0                  1910.49   
      43386009019                    169.0                  2137.42   
      43386009019                    144.0                  1920.29   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
FL    43386009019                     1906.79                             3.7   
      43386009019                     2137.42                             0.0   
      43386009019                     1920.29                             0.0   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
FL    43386009019  (27.8333, -81.717)                   3808334  
      43386009019  (27.8333, -81.717)                   3808334  
      43386009019  (27.8333, -81.717)                   3808334

In [73]:
medicaid_data_w_enrollments_hierarch.loc[('OR', 43386009019)]

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
OR    43386009019             FFSU   GaviLyte -           84000.0   
      43386009019             MCOU   GaviLyte -         1240000.0   
      43386009019             FFSU   GaviLyte -           72000.0   
      43386009019             FFSU   GaviLyte -           52000.0   
      43386009019             MCOU   GaviLyte -         1960000.0   
      43386009019             MCOU   GaviLyte -          780000.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
OR    43386009019                     18.0                   403.80   
      43386009019                    294.0                  4129.53   
      43386009019                     17.0                   314.64   
      43386009019                     13.0                   230.30   
      43386009019                    467.0                  6145.37   
      43386009019                    188.0                  2533.08   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
OR    43386009019                      395.51                            8.29   
      43386009019                     4129.53                            0.00   
      43386009019                      312.60                            2.04   
      43386009019                      203.36                           26.94   
      43386009019                     6145.37                            0.00   
      43386009019                     2533.08                            0.00   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
OR    43386009019  (44.5672, -122.1269)                   1123913  
      43386009019  (44.5672, -122.1269)                   1123913  
      43386009019  (44.5672, -122.1269)                   1123913  
      43386009019  (44.5672, -122.1269)                   1123913  
      43386009019  (44.5672, -122.1269)                   1123913  
      43386009019  (44.5672, -122.1269)                   1123913

In [74]:
medicaid_data_w_enrollments_hierarch.loc[('WA', 43386009019)]

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
WA    43386009019             FFSU   GaviLyte -            58.000   
      43386009019             FFSU   GaviLyte -            65.000   
      43386009019             MCOU   GaviLyte -          2702.118   
      43386009019             MCOU   GaviLyte -          2642.000   
      43386009019             MCOU   GaviLyte -          2753.149   
      43386009019             FFSU   GaviLyte -            87.000   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
WA    43386009019                     56.0                  1083.76   
      43386009019                     65.0                  1045.00   
      43386009019                   2677.0                 31709.24   
      43386009019                   2578.0                 28734.28   
      43386009019                   2719.0                 34745.72   
      43386009019                     85.0                  1389.96   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
WA    43386009019                     1083.76                            0.00   
      43386009019                     1031.66                           13.34   
      43386009019                    31691.66                           17.58   
      43386009019                    28732.78                            1.50   
      43386009019                    34743.22                            2.50   
      43386009019                     1387.77                            2.19   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
WA    43386009019  (47.3917, -121.5708)                   1771679  
      43386009019  (47.3917, -121.5708)                   1771679  
      43386009019  (47.3917, -121.5708)                   1771679  
      43386009019  (47.3917, -121.5708)                   1771679  
      43386009019  (47.3917, -121.5708)                   1771679  
      43386009019  (47.3917, -121.5708)                   1771679

* Find and list all unique `NDC`s for which the difference between the largest and second large log-scale ratio-by-state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ratio of  `-12.623428`

In [96]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc

NDC    2143380      2143480      2144511      2144527      2197590      \
STATE                                                                    
AK       -9.609109   -10.008280          NaN          NaN          NaN   
AL       -9.940595    -9.805485          NaN          NaN          NaN   
AR      -12.985157   -12.657103          NaN          NaN          NaN   
AZ      -11.533870   -10.821194          NaN          NaN          NaN   
CA      -10.926863   -10.698372   -16.127018          NaN    -9.464364   
CO      -10.862982    -9.925954          NaN          NaN    -7.608472   
CT       -8.405103    -7.948475   -12.411014          NaN    -6.677343   
DE             NaN   -10.274536          NaN          NaN          NaN   
FL      -13.534299   -12.569558   -18.053374          NaN   -10.239592   
GA      -15.018034   -13.858835          NaN          NaN          NaN   
HI       -8.747994    -7.987181          NaN          NaN          NaN   
IA             NaN   -11.726673          NaN          NaN          NaN   
ID      -10.620398   -10.014219          NaN          NaN          NaN   
IL      -10.688482   -10.406940   -16.181427          NaN          NaN   
IN       -8.305096    -7.688026          NaN          NaN    -7.584248   
KS      -10.895518    -9.835536          NaN          NaN    -7.487729   
KY       -8.325745    -8.097982   -14.338234          NaN    -6.942597   
LA       -9.150146    -8.747732          NaN          NaN          NaN   
MA       -7.921997    -7.803463   -13.472194          NaN    -7.741402   
MD       -8.788942    -8.598997          NaN          NaN    -6.491564   
ME      -10.522622    -9.141564          NaN          NaN          NaN   
MI      -12.343430   -11.316613          NaN          NaN          NaN   
MN       -8.661399    -7.883399   -14.481856          NaN          NaN   
MO      -11.183484   -10.206768          NaN          NaN    -9.549612   
MS      -13.744104          NaN          NaN          NaN          NaN   
MT      -10.072184    -9.539963          NaN          NaN          NaN   
NC      -10.721813   -10.172403   -15.234263          NaN          NaN   
ND             NaN          NaN          NaN          NaN          NaN   
NE      -10.225587   -10.011462          NaN          NaN          NaN   
NH      -11.338110   -10.425886          NaN          NaN          NaN   
NJ      -11.063407   -10.389527   -16.379915          NaN    -8.537564   
NM      -11.101464   -11.164200          NaN          NaN          NaN   
NV       -9.527766    -9.015820          NaN          NaN          NaN   
NY       -9.660599    -9.235781   -14.055095          NaN    -6.634022   
OH       -8.540125    -8.629338   -14.069577    -16.63842    -5.551620   
OK      -12.358208   -10.550853          NaN          NaN          NaN   
OR      -11.467104   -11.061180          NaN          NaN          NaN   
PA       -7.378007    -7.672800   -14.663549          NaN    -8.515650   
RI      -10.377028   -12.280813          NaN          NaN          NaN   
SC      -10.695086   -10.028451          NaN          NaN    -9.098535   
SD       -9.568199    -9.218392          NaN          NaN          NaN   
TN      -12.240261   -11.699134          NaN          NaN          NaN   
TX      -11.361813   -10.857851          NaN          NaN          NaN   
UT      -12.066772   -10.625645          NaN          NaN          NaN   
VA       -9.779228    -9.485665          NaN          NaN          NaN   
VT      -11.085710   -10.407638          NaN          NaN          NaN   
WA      -11.374062   -11.446073   -14.825948          NaN          NaN   
WI      -14.299344   -14.884306          NaN          NaN          NaN   
WV      -10.526494    -9.958528          NaN          NaN          NaN   
WY             NaN          NaN          NaN          NaN          NaN   

NDC    2322730      2322830      2322930      2323030      2323130      ...  \
STATE                                                                   

In [106]:
twolargest = medicaid_norm_ndc.apply(lambda x: x.nlargest(2), axis=0)
difference = twolargest.apply(lambda x: x.cummax()-x.cummin(), axis=0)
final_list = difference.stack()[abs(difference.stack())>10]
final_list

STATE  NDC        
CA     641047721      10.587633
       39822420006    12.484298
IN     409710002      12.040548
LA     50458016601    10.646297
MO     9023303        11.633238
       407141233      12.441317
MT     406036501      10.957791
NE     409796709      11.582770
NV     409767009      15.910902
OK     574082710      11.549606
       591289749      11.597786
       60793060002    12.023433
       68001028529    10.396696
SD     63323017530    11.659012
TX     264761210      13.096539
VA     68180062210    16.514544
WI     338007304      14.209622
dtype: float64

* The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states
   * Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   * The second highest state has a log value of -12.623428 (0.000158478197834722)
* Oklahoma is not a high-incidence state for cancer
* Could the following explain what is happening in Oklahoma?
  * The following is clinical trial conducted by the University of Oklahoma

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10